src/core/features/tags/controller/TagsController.ts

```diff
@@ -55,24 +125,89 @@ export class TagsController {
 	}
 
 	public async add(name: string, parent: null | string): Promise<string> {
-		const db = wrapDB(this.db.get());
-
-		const segments = name.split('/');
+		validateTagName(name);
 
 		let lastId: string | null = null;
-		if (segments.length > 1) {
-			await this.db.get().transaction(async (tx) => {
-				const db = wrapDB(tx);
 
-				for (let idx = 0; idx < segments.length; idx++) {
-					const name = segments[idx];
+		await this.db.get().transaction(async (tx) => {
+			const db = wrapDB(tx);
+
+			let resolvedTagName: string;
+			if (!parent) {
+				resolvedTagName = name;
+			} else {
+				const {
+					rows: [parentTag],
+				} = await db.query(
+					selectResolvedTags(this.workspace, {
+						where: [qb.sql`id = ${parent}`],
+						limit: 1,
+					}),
+					RowScheme,
+				);
+				// If the parent tag is not found in the database, the tag cannot be created
+				if (!parentTag)
+					throw new TagControllerError(
+						`Parent tag ${parent} does not exist`,
+						TAG_ERROR_CODE.PARENT_TAG_NOT_EXIST,
+					);
+				resolvedTagName = `${parentTag.resolvedName}/${name}`;
+			}
+
+			// Check tag uniqueness
+			const {
+				rows: [duplicateTag],
+			} = await db.query(
+				selectResolvedTags(this.workspace, {
+					where: [qb.sql`resolved_name = ${resolvedTagName}`],
+					limit: 1,
+				}),
+				RowScheme,
+			);
+			if (duplicateTag) {
+				throw new TagControllerError(
+					`Tag ${duplicateTag.resolvedName} already exists`,
+					TAG_ERROR_CODE.DUPLICATE,
+				);
+			}
+
+			const resolvedTagSegments = resolvedTagName.split('/');
+			if (resolvedTagSegments.length > 1) {
+				// Build an array of all path variants to find existing tags that match exactly this one
+				// 'foo/bar/baz' - ['foo', 'foo/bar', 'foo/bar/baz']
+				// The found tag allows detecting non-existing segments from resolvedTagSegments and creating only those
+				const tagNameVariants = resolvedTagSegments.reduce<string[]>(
+					(prev, segment) => [
+						...prev,
+						prev.length ? `${prev[prev.length - 1]}/${segment}` : segment,
+					],
+					[],
+				);
+				const {
+					rows: [rootTag],
+				} = await db.query(
+					selectResolvedTags(this.workspace, {
+						where: [qb.sql`resolved_name IN (${qb.values(tagNameVariants)})`],
+						order: qb.sql`ORDER BY LENGTH(resolved_name) DESC`,
+						limit: 1,
+					}),
+					RowScheme,
+				);
+
+				const parentTagId = rootTag ? rootTag.id : parent;
```

Comment:
Why we use a parent tag in case we can't find any root tag? We should run special logic then to create root tag.

And we shoud add tests for that case. If you can't reproduce this case we could throw error here